### PR TITLE
Add Rust bindings for LSIF typed

### PR DESCRIPTION
**Pull request by olafurpg**
Requires us to run `cargo install protobuf-codegen` before invoking `buf generate`. We should encode these steps in a build script somewhere.

### CODE REVIEW

**Review by efritz — APPROVED** [review]

**Comment by efritz** [comment]
@olafurpg This also presumably needs rust installed on the buildkite agent?

**Comment by varungandhi-src** [comment]
> Requires us to run cargo install protobuf-codegen before invoking buf generate. We should encode these steps in a build script somewhere.

Couple of things that would be useful:
1. Add a CI step to check that the generated source is up-to-date with the schema.
2. Document this in the README as well (or point the README to the CI step).

**Comment by olafurpg** [comment]
@varungandhi-src I updated the `./dev/proto-generate.sh` script to install `cargo` and `protoc-gen-rust` in CI. If cargo is already installed for some reason then we can drop the first curl command.

**Comment by olafurpg** [comment]
CI failed with
```
Rust is installed now. Great!
error: no override and no default toolchain set
proto/gen.go:3: running "../dev/proto-generate.sh": exit status 1
````
@varungandhi-src any advice on how to set the default toolchain if it's missing? I prefer not to invoke `rustup default stable` on everybody's computers if they have customized their environment to use a different toolchain

**Comment by varungandhi-src** [comment]
@olafurpg you should be able to use a [toolchain file](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file). Since the error says:

> error: no override and no default toolchain set

The toolchain file will create a local override without having to set a global default.

File: dev/proto-generate.sh [hunk 1]
```diff
@@ -6,7 +6,9 @@ cd "$(dirname "${BASH_SOURCE[0]}")/.." # cd to repo root dir
 
 echo "--- cargo install rust-protobuf"
 curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --default-toolchain none -y
```
Comment (efritz) [inline dev/proto-generate.sh]: I'm not sure we'll want to do the installation step here. We should instead have rust/cargo installed via asdf on the buildkite-agent itself: https://github.com/sourcegraph/infrastructure/tree/main/docker-images/buildkite-agent

**Comment by olafurpg** [comment]
The CI should be failing here. Will need to investigate better

**Comment by olafurpg** [comment]
The CI is failing because `cargo` is missing. https://buildkite.com/sourcegraph/sourcegraph/builds/126210#c2e2dd98-5c43-4acb-a452-280dc7983b49/947

 There exists a Rust asdf plugin that installs `cargo` but I didn't figure out how to update the build-kite image to install it, there was no place to insert the asdf plugin URL

**Comment by efritz** [comment]
@olafurpg Add your plugin to [this list](https://github.com/sourcegraph/infrastructure/blob/main/docker-images/buildkite-agent/add-asdf-plugins.sh#L5) and follow [these instructions](https://github.com/sourcegraph/infrastructure/tree/main/docker-images/buildkite-agent) to deploy a new agent.

**Comment by efritz** [comment]
@olafurpg Did this for you in https://github.com/sourcegraph/infrastructure/pull/3053.

**Comment by tjdevries** [comment]
Closed in favor of: https://github.com/sourcegraph/sourcegraph/pull/30016 since I want to put it in it's own crate for re-use outside of SG if necessary.

**Pull request closed without merging**